src/constraint/4C_constraint_springdashpot.cpp

```diff
@@ -10,21 +10,228 @@
 #include "4C_adapter_coupling_nonlin_mortar.hpp"
 #include "4C_contact_interface.hpp"
 #include "4C_fem_condition_utils.hpp"
+#include "4C_fem_general_cell_type.hpp"
+#include "4C_fem_general_cell_type_traits.hpp"
+#include "4C_fem_general_element_integration.hpp"
+#include "4C_fem_general_element_integration_select.hpp"
+#include "4C_fem_general_extract_values.hpp"
 #include "4C_fem_general_node.hpp"
 #include "4C_global_data.hpp"
 #include "4C_io.hpp"
 #include "4C_io_pstream.hpp"  // has to go before io.hpp
+#include "4C_linalg_fixedsizematrix.hpp"
 #include "4C_linalg_utils_sparse_algebra_assemble.hpp"
 #include "4C_linalg_utils_sparse_algebra_create.hpp"
+#include "4C_linalg_utils_sparse_algebra_manipulation.hpp"
 #include "4C_truss3.hpp"
+#include "4C_utils_exceptions.hpp"
 #include "4C_utils_function.hpp"
 #include "4C_utils_function_of_time.hpp"
 
-#include <iostream>
+#include <algorithm>
+#include <concepts>
+#include <tuple>
 #include <utility>
 
 FOUR_C_NAMESPACE_OPEN
 
+namespace
+{
+
+  double scale_by_optional_function(double baseline_value, int num_func, double total_time)
+  {
+    return num_func ? Global::Problem::instance()
+                              ->function_by_id<Core::Utils::FunctionOfTime>(num_func)
+                              .evaluate(total_time) *
+                          baseline_value
+                    : baseline_value;
+  }
+
+  template <typename T>
+  concept SpringDashpotEvaluatable = requires(T t, double delta_displacement, double velocity) {
+    { t(delta_displacement, velocity) } -> std::convertible_to<std::tuple<double, double, double>>;
+  };
+
+  template <SpringDashpotEvaluatable SpringDashpot, Core::FE::CellType celltype, unsigned int dim>
+  void add_robin_spring_dashpot_contribution_xyz(const Core::Elements::Element& surface_element,
+      const Core::Elements::ElementNodes<celltype, dim> element_nodes,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& displacements,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& displacement_offset,
+      const std::vector<double>& constant_offset,
+      const Core::LinAlg::Matrix<Core::FE::num_nodes<celltype>, dim>& velocity,
+      const std::vector<SpringDashpot>& spring_dashpot_evaluator, const double time_factor,
+      const Core::FE::GaussIntegration& gauss_integration, const std::vector<bool>& onoff,
+      std::optional<Core::LinAlg::Matrix<dim * Core::FE::num_nodes<celltype>, 1>>& residual_vector,
+      std::optional<Core::LinAlg::Matrix<dim * Core::FE::num_nodes<celltype>,
+          dim * Core::FE::num_nodes<celltype>>>& stiffness_matrix)
```

Comment:
You might want to assert that at least one of these is non-empty, so the function call has any effect.